product_state/models/product_template.py

```diff
@@ -1,18 +1,74 @@
 # Copyright 2017 ACSONE SA/NV (<http://acsone.eu>)
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
 
-from odoo import models, fields
+from odoo import api, models, fields
+
+
+class ProductState(models.Model):
+    _name = 'product.state'
+    _description = "Product State"
+    _order = 'sequence, id'
+
+    name = fields.Char('State Name', required=True, translate=True)
+    code = fields.Char('State Code', required=True)
+    sequence = fields.Integer('Sequence', help="Used to order the States", default=25)
+    description = fields.Text(translate=True)
+    product_ids = fields.One2many(
+        'product.template',
+        'product_state_id',
+        string='State Products',
+    )
+    products_count = fields.Integer(
+        string='Number of products',
+        compute='_compute_products_count',
+    )
+
+    _sql_constraints = [
+        ('code_unique', 'UNIQUE(code)', 'Product State Code must be unique.')
+        ]
+
+    @api.depends('product_ids')
+    def _compute_products_count(self):
+        data = self.env["product.template"].read_group([
+            ('product_state_id', 'in', self.ids),
+        ], ['product_state_id'], ['product_state_id'])
+        mapped_data = {
+            record['product_state_id'][0]: record['product_state_id_count']
+            for record in data
+        }
+        for state in self:
+            state.products_count = mapped_data.get(state.id, 0)
 
 
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
-    state = fields.Selection(selection=[
-        ('draft', 'In Development'),
-        ('sellable', 'Normal'),
-        ('end', 'End of Lifecycle'),
-        ('obsolete', 'Obsolete')],
-        string='Status',
-        default='sellable',
-        index=True
+    state = fields.Char(
+        string='Product Status',
+        index=True,
+    )
+    product_state_id = fields.Many2one(
+        'product.state',
+        string='State',
+        help='Select a state for this product',
+        compute="_compute_product_state",
+        inverse="_inverse_product_state",
+        store=True,
     )
+
+    @api.depends('state')
+    def _compute_product_state(self):
```

Comment:
@emagdalenaC2i I don't think this is the good manner doing what you want. The code seems weird to me.

I would have done the inverse. Set the 'state' field as a computed stored one and the state_id as normal many2one.

@Cedric-Pigeon do you have an advice on this ?

Reply:
ping @kos94ok-3D What do you think?

Reply:
I don't mind.

Reply:
I tend to agree with @rousseldenis - that way the new "work" field is dreict write, and not hindered by computations. The "old" field is there for backward compatibility, is the computed one, and eventually will go away in future versions.

One more note is that "state" is usually a closed selection field, and for dynamic many2one the usual name is "stage_id". It would be nice to stick to the common field naming.